### ISSUE TITLE
Module proposal: Performance-oriented indexing for WordPress database tables

### ISSUE THREAD

**Issue by OllieJones** [WordPress/performance#338]
## Feature Description
This proposes a **module** for the **Performance Lab Plugin**. It's intended to follow the process for [Proposing a new module](https://github.com/WordPress/performance/blob/add/new-module-doc/docs/Proposing-a-new-module.md).  

See also #132.

The indexes on WordPress tables need improvement, especially to handle sites with many posts or users. Many WooCommerce sites, especially those with many products, especially need improved indexes. The improved indexes affect many queries generated by WordPress core and by plugins and themes. They allow the database server to avoid scanning through tables to retrieve just a few rows of data. It's common industry knowledge among SQL programmers and database administrators that indexes can improve query response times dramatically when properly designed. Faster database operations **save power in data centers** and **improve audience satisfaction**.

The proposed module alters tables' indexes and then gets out of the way. It doesn't need to hook anything. 

* The opportunity for improving the indexes is partly the result of continued improvements in the MariaDB and MySQL database software products used by WordPress site operators. In the early days of WordPress, MySQL only offered the MyISAM storage engine, and all WordPress sites used it. Later, MySQL and its MariaDB fork added the InnoDB storage engine with the COMPACT row format (known as Antelope). Most recently, since MySQL 5.7 and MariaDB 10.2, the InnoDB storage engine with the DYNAMIC row format has become available (known as Barracuda).

 * The opportunity also comes from the organization of InnoDB tables. In their *clustered index* organization, the data of each table is stored with the table's primary key. So, database lookups using a table's primary key (for example, `WHERE option_id = 42`) are faster than lookups with some other indexed column (for example `WHERE option_name = 'answer_to_everything'`.  

   Several WordPress tables, including the wp_options table, have primary keys on columns that are rarely used for lookups. So part of this proposal involves altering those tables' primary keys so they match the most common lookup patterns.  In this example, making the `option_name` column the table's primary key improves query response time.

This proposal is based a year's experience with the non-monetized [Index WP MySQL For Speed](https://wordpress.org/plugins/index-wp-mysql-for-speed/) plugin (source code [here](https://github.com/OllieJones/index-wp-mysql-for-speed)). It's also based on the longer experience of that plugin's authors, @rjasdf and @OllieJones, doing various one-off  indexing changes for various WordPress site operators.

### What does this module do?

> Note: throughout this document tables are named with the default `wp_` prefix, even though some tables use different prefixes.

It updates tables' storage engines, when necessary, from the older MyISAM storage engine to the more capable InnoDB engine and DYNAMIC storage format. 

It  alters the indexes on eight database tables in WordPress's core.:

* wp_comments
* wp_commentmeta
* wp_options
* wp_posts
* wp_postmeta
* wp_termmeta
* wp_users
* wp_usermeta

The proposed changes to indexes are [here](https://www.plumislandmedia.net/index-wp-mysql-for-speed/tables_and_keys/).  The actual data-definition-language SQL statements to make the changes are shown [here for Antelope databases](https://www.plumislandmedia.net/wordpress/index-wp-mysql-for-speed-ddl-antelope/) and here for [Barracuda databases](https://www.plumislandmedia.net/wordpress/index-wp-mysql-for-speed-ddl-antelope/). 

(Notice that the proof-of-concept plugin's code is organized to change only the indexes and tables needing changing.)

### Basic Information

* Proposed Slug:  `database/performance-indexes`
* Proposed Name:  Index database for performance (under the Database heading)
* Proposed Ownership:  @OllieJones   (volunteer dev) 
 
### Decisions to make when building this module into performance-lab.

* Unlike other modules, this one needs at least some dashboard user interface, to initiate the reindexing of the tables, and to complain if the tables use legacy storage features (MyISAM, COMPACT row format). **Where do we put that user interface panel?**
* Experience shows that the POC plugin's wp-cli access is necessary on large sites to do the reindexing without timeouts. **Should performance lab get wp-cli access? How about using WP_Cron?**
*  **What should the module do on multisite installations?**

### Possible complications in the real world

(Lots of deep-dive details here.)

* The proof-of-concept (POC) plugin offers to convert MyISAM tables to InnoDB tables, and does not offer to reindex MyISAM tables. That is by design: the plugin's indexes rely on InnoDB features for their performance.
* To support the [most sites](https://wordpress.org/about/stats/) it's necessary to handle both Antelope (MySQL 5.5 and 5.6) and Barracuda storage engines.  (The difference relates to [prefix indexes](https://www.plumislandmedia.net/index-wp-mysql-for-speed/wordpresss-prefix-keys/).)  The majority of WordPress installations still use Antelope, due to hosting services not upgrading. This support generates a lot of complexity in the POC plugin. 
* Experience shows some users want to revert the indexes to standard.
* Database upgrades that come with some WordPress core upgrades can change indexes. 
* Some sites already have customized indexes. The proof-of-concept plugin has logic to convert those indexes to the ones proposed. That, too, is complex.
* Some plugins create their own indexes for their own purposes. For example, WooCommerce puts its own index on wp_comments. ([See here](https://wordpress.org/support/topic/update-to-6-4-problematic-db-update-2/).)  A blocklist of index-name prefixes helps. [Another plugin also creates indexes](https://github.com/OllieJones/index-wp-mysql-for-speed/issues/35). 
* Indexes consume SSD / hard-disk space on database servers. Some legacy sites may not have enough space, and there's no reliable way for WordPress code (be it module, plugin, or core) to determine how much space is available, until an index-conversion attempt throws an error.
* Index conversion takes time for large tables.  Some php / webserver configurations time out too soon. The POC plugin supports wp-cli to help site operators get around this.
* There's a design tradeoff in indexing `wp_options`. The POC plugin chose to optimize for the case where the site operator uses a [persistent object cache](https://developer.wordpress.org/reference/classes/wp_object_cache/#persistent-caching), instead of optimizing for retrieving autoloaded options.
* Multisite: the POC plugin works on multisite WordPress instances, but it is organized (as of v 1.4.4) to be run separately for each site. A decision about how to handle multisite setups is necessary.

**Comment by tillkruss** [comment]
Related #132.

**Comment by pbearne** [comment]
Will the #347 proposal help with the trade-off "There's a design tradeoff in indexing wp_options. The POC plugin chose to optimize for the case where the site operator uses a [persistent object cache](https://developer.wordpress.org/reference/classes/wp_object_cache/#persistent-caching), instead of optimizing for retrieving autoloaded options"?

**Comment by OllieJones** [comment]
@pbearne  I don't think so, at least not much. The choices in the tradeoff were:

* Use `(autoload, option_name)`  as the primary key (PK). That way `SELECT * FROM wp_options WHERE autoload='yes'` uses the PK's clustered index. Without a persistent object cache that query runs once on every pageload and cranks out a lot of data. So streamlining it pays off. With any object caching scheme in place that query is far far less frequent.
* Use `(option_name)` as the PK. That way queries (SELECT, UPDATE, DELETE)  containing  `WHERE option_name = 'whatever'` filter clauses can use the clustered index. There are lots of these in typical pageviews.

   Of course, stuff works fine with a non-primary UNIQUE key on `(option_name)` . Using the clusted index, it's just a little more streamlined, and that's the point.

So, with the assumption that the biggest sites will have the best infrastructure (meaning redis or memcached) I went for the second alternative. I don't think #347 will use that table in significantly different ways than what I've outline. (I could be wrong of course).

**Comment by rjasdf** [comment]
@pbearne - (Adding on to what Ollie said.)  I'll split the first bullet into two cases:

If `autoload` is `yes` for most rows, `PRIMARY KEY(option_name)` is nearly as fast for `SELECT * FROM wp_options WHERE autoload='yes'` as when using the 2-column PK.  This is because it will simply read the entire table, ignoring any `no` entries.

If autoload is `no` for most rows, then caching isn't doing much; hence, they benefit of the two-column PK is minor.

Bottom line:  `PRIMARY KEY(option_name)`.  In some situations, `INDEX(autoload)` might be worth having.

**Comment by OllieJones** [comment]
**TL;DR**  This proposed module doesn't seem right for the performance lab plugin.

The #core-performance group offered a great discussion of this proposal. [Summary here](https://make.wordpress.org/core/2022/06/28/performance-team-meeting-summary-28-june-2022/).

The reasons to omit the reindexing from a performance lab module:

* doing it successfully on large tables can require wp-cli commands to avoid timeouts. Performance lab doesn't contemplate providing wp-cli interfaces for its modules.
* the reindexing probably requires an additional dashboard panel that's possibly fairly complex.  [Here's a description](https://www.plumislandmedia.net/index-wp-mysql-for-speed/high_performance_keys/) of what might need to go in that panel from the proof-of-concept plugin.  While implementing the panel is not a problem, the additional site-owner complexity might be.

**One suggestion**: create a standalone wp-cli module to generate the SQL Data Definition Language statements to do the reindexing, and to run those statements. 

**Another suggestion**:  rather than putting this module into the performance lab plugin, break it up into...

1. Code in core's schema.php to use the proposed new table indexes in *new installs* of WordPress. This might be done soon.
2. Code in schema.php to convert tables to new indexes on version update. This can be in a later version.

These suggestions will make for added complexity in schema.php. Why?

The exact indexes used are dependent on the MariaDB / MySQL server version in use. Older versions, with the InnoDB Antelope storage engine, need [prefix keys](https://www.plumislandmedia.net/index-wp-mysql-for-speed/wordpresss-prefix-keys/).  

So, schema.php upgrades will have to deal with two version-dependent schema versions and a total of six use cases.

1. Antelope (prefix-requiring) "legacy" indexing.
2. Antelope new indexing on new installs.
3. Antelope reindexing on updates.
4. Barracuda (no prefixes) "legacy" indexing.
5. Barracuda new indexing on new installs.
6. Barracuda reindexing on updates.

That's a lot of use cases. Keeping them straight in the field will be challenging, both for install/update, for testing, and for troubleshooting. I, @OllieJones, am reluctant to put all this reindexing in core for that reason. (Please correct me if I'm wrong about this.)

So, it looks to me like a performance lab module isn't the right way to deploy these optimizations.

I'm updating the Index WP MySQL for Speed plugin soon to allow this wp-cli construct. It gets the plugin to generate the appropriate SQL statements then uses `wp db query` to run them. And, of course the statements can be displayed or put into a file, UNIX style.

```bash
wp index-mysql enable --all --dryrun | wp db query
```

**Comment by mxbclang** [comment]
@OllieJones Thank you so much for this detailed summary and for taking the time to discuss this with the team! 

Per your note, we won't proceed with this proposal as a module for the Performance Lab plugin for now, but look forward to seeing it elsewhere in the future. I'll mark this issue as closed.
